pkg/asset/store.go

```diff
@@ -86,47 +122,128 @@ func (s *StoreImpl) Save(dir string) error {
+// dependencies if necessary.
+// It returns dirty if the asset or any of its parents is loaded from on-disk files.
+func (s *StoreImpl) fetch(asset Asset, indent string) (dirty bool, err error) {
 	logrus.Debugf("%sFetching %s...", indent, asset.Name())
+
+	// Return immediately if the asset is found in the cache,
+	// this is because we are doing a depth-first-search, it's guaranteed
+	// that we always fetch the parent before children, so we don't need
+	// to worry about invalidating anything in the cache.
 	storedAsset, ok := s.assets[reflect.TypeOf(asset)]
 	if ok {
 		logrus.Debugf("%sFound %s...", indent, asset.Name())
-		reflect.ValueOf(asset).Elem().Set(reflect.ValueOf(storedAsset).Elem())
-		return nil
+		reflect.ValueOf(asset).Elem().Set(reflect.ValueOf(storedAsset.asset).Elem())
+		return storedAsset.dirty, nil
 	}
 
 	dependencies := asset.Dependencies()
 	parents := make(Parents, len(dependencies))
 	if len(dependencies) > 0 {
 		logrus.Debugf("%sGenerating dependencies of %s...", indent, asset.Name())
 	}
+
+	var anyParentsDirty bool
 	for _, d := range dependencies {
-		err := s.fetch(d, indent+"  ")
+		dt, err := s.fetch(d, indent+"  ")
 		if err != nil {
-			return errors.Wrapf(err, "failed to fetch dependency for %s", asset.Name())
+			return false, errors.Wrapf(err, "failed to fetch dependency for %s", asset.Name())
+		}
+		if dt {
+			anyParentsDirty = true
 		}
 		parents.Add(d)
 	}
 
-	// Before generating the asset, look if we have it all ready in the state file
-	// if yes, then use it instead
+	// Try to load from on-disk files first.
+	var foundOnDisk bool
+	as, ok := asset.(LoadableAsset)
+	if ok {
+		logrus.Debugf("%sLoading asset %s from disk", indent, asset.Name())
+		foundOnDisk, err = as.Load(s.fileFetcher)
+		if err != nil {
+			return false, errors.Wrapf(err, "unexpected error when loading asset %s", asset.Name())
+		}
+		if foundOnDisk {
+			logrus.Debugf("%sFound %s on disk...", indent, asset.Name())
+			wa, ok := asset.(WritableAsset)
+			if !ok {
+				return false, errors.Errorf("on-disk asset %q must be writable asset", asset.Name())
+			}
+			s.onDiskAssets = append(s.onDiskAssets, wa)
+		}
+	}
+
+	// Try to load from the state file.
 	logrus.Debugf("%sLooking up asset from state file: %s", indent, reflect.TypeOf(asset).String())
-	ok, err := s.GetStateAsset(asset)
+	foundInStateFile, err := s.GetStateAsset(asset)
 	if err != nil {
-		return errors.Wrapf(err, "failed to unmarshal asset %q from state file %q", asset.Name(), stateFileName)
+		return false, errors.Wrapf(err, "failed to unmarshal asset %q from state file %q", asset.Name(), stateFileName)
 	}
-	if ok {
-		logrus.Debugf("%sAsset found in state file", indent)
-	} else {
-		logrus.Debugf("%sAsset not found in state file. Generating %s...", indent, asset.Name())
-		err := asset.Generate(parents)
-		if err != nil {
-			return errors.Wrapf(err, "failed to generate asset %s", asset.Name())
+
+	dirty = anyParentsDirty || foundOnDisk
```

Comment:
Better idea that we check dirty first and then only if its false should we look up the state file. If the asset or parents are dirty, then we don't care about state file anyway (at least for now until we want conflict resolution).

```
        foundInStateFile := false
        dirty = anyParentsDirty || foundOnDisk
        if !dirty {
	    foundInStateFile, err := s.GetStateAsset(asset)
	    if err != nil {
	          return false, errors.Wrapf(err, "failed to unmarshal asset %q from state file %q", asset.Name(), stateFileName)
	    }
	}
```

Reply:
The check for the state file is because we want to print some warning message if the object is both on disk and in state file.